brave/src/main/java/brave/propagation/CurrentTraceContext.java

```diff
@@ -21,23 +24,82 @@
  * com.google.inject.servlet.RequestScoper and com.github.kristofa.brave.CurrentSpan
  */
 public abstract class CurrentTraceContext {
+
+  /** Implementations of this allow standardized configuration, for example scope decoration. */
+  public abstract static class Builder<B extends Builder<B>> {
+    ArrayList<ScopeDecorator> scopeDecorators = new ArrayList<>();
+
+    /** Implementations call decorators in order to add features like log correlation to a scope. */
+    public B addScopeDecorator(ScopeDecorator scopeDecorator) {
+      if (scopeDecorator == null) throw new NullPointerException("scopeDecorator == null");
+      this.scopeDecorators.add(scopeDecorator);
+      return (B) this;
+    }
+
+    public abstract CurrentTraceContext build();
+  }
+
   /** Returns the current span in scope or null if there isn't one. */
   public abstract @Nullable TraceContext get();
 
   /**
-   * Sets the current span in scope until the returned object is closed. It is a programming
-   * error to drop or never close the result. Using try-with-resources is preferred for this reason.
+   * Sets the current span in scope until the returned object is closed. It is a programming error
+   * to drop or never close the result. Using try-with-resources is preferred for this reason.
    *
    * @param currentSpan span to place into scope or null to clear the scope
    */
   public abstract Scope newScope(@Nullable TraceContext currentSpan);
 
+  final List<ScopeDecorator> scopeDecorators;
+
+  public interface Factory {
+    CurrentTraceContext create(List<ScopeDecorator> scopeDecorators);
+  }
+
+  protected CurrentTraceContext() {
+    this.scopeDecorators = Collections.emptyList();
+  }
+
+  protected CurrentTraceContext(Builder<?> builder) {
+    this.scopeDecorators = (List<ScopeDecorator>) builder.scopeDecorators.clone();
+  }
+
+  /**
+   * When implementing {@linkplain #newScope(TraceContext)}, decorate the result before returning
+   * it.
+   *
+   * <p>Ex.
+   * <pre>{@code
+   *   @Override public Scope newScope(@Nullable TraceContext currentSpan) {
+   *     final TraceContext previous = local.get();
+   *     local.set(currentSpan);
+   *     class ThreadLocalScope implements Scope {
+   *       @Override public void close() {
+   *         local.set(previous);
+   *       }
+   *     }
+   *     Scope result = new ThreadLocalScope();
+   *     // ensure scope hooks are attached to the result
+   *     return decorateScope(currentSpan, result);
+   *   }
+   * }</pre>
+   */
+  protected Scope decorateScope(@Nullable TraceContext currentSpan, Scope scope) {
+    int length = scopeDecorators.size();
+    if (length == 0) return scope;
```

Comment:
I think the for loop duplicates this check so it doesn't really speed up the `0` case.

Reply:
agreed